File: prime-csharp/src/Prime.cs
using System.Collections.Generic;

namespace Prime;

public static class Prime {
    
    public static bool IsPrime(uint n) {
        for (uint i = 2; i <= n / 2; ++i) {
            if (n % i == 0)
                return false;
        }
        return true;
    }

    public static uint CountPrimes(uint upTo, uint numberOfThreads) {
        List<Thread> threads = new();
        uint numberOfPrimeNumbers = 0;
        uint currentNumber = 2;
        for (uint t = 0; t < numberOfThreads; t++) {
            var thread = new Thread(() => {
                while(true){
                    uint localCurrentNumber = Interlocked.Increment(ref currentNumber) - 1;  
                    if(localCurrentNumber >= upTo) {
                        break;
                    }
                    if (IsPrime(localCurrentNumber)) {
                        Interlocked.Increment(ref numberOfPrimeNumbers); 
                    }   
                }
            });    
            threads.Add(thread);
            thread.Start();
        }
        threads.ForEach(t => t.Join());
        return Interlocked.Exchange(ref numberOfPrimeNumbers, 0);
    }

}

File: prime-kotlin/app/src/main/kotlin/prime/kotlin/Prime.kt
package prime.kotlin

import kotlin.concurrent.thread
import java.util.concurrent.atomic.AtomicInteger

fun isPrime(n: UInt): Boolean {
    val end = (n / 2u)
    for (i in 2u..end) {
        if ((n % i) == 0u) {
            return false
        }
    }
    return true    
}

fun countPrimes(upTo: UInt, numberOfThreads: UInt): UInt {
    var threads = mutableListOf<Thread>()
    var numberOfPrimeNumbers = AtomicInteger(0)
    var currentNumber =  AtomicInteger(2)
    
    for (t in 0u..< numberOfThreads) {
        threads.add(thread {
            while(true){
                var localCurrentNumber = currentNumber.getAndAdd(1).toUInt();
                if(localCurrentNumber >= upTo) {
                    break;
                }
                if (isPrime(localCurrentNumber.toUInt())) {
                    numberOfPrimeNumbers.incrementAndGet()
                }    
            }
        });    
    }    

    for (t in threads) {
        t.join();
    }

    return numberOfPrimeNumbers.get().toUInt()
}